extensions/stackdriver/edges/edge_reporter.cc

```diff
@@ -0,0 +1,149 @@
+      local_node_info.platform_metadata().at(Common::kGCPProjectKey);
+  current_request_->set_parent("projects/" + project_id);
+
+  // TODO(dougreid): figure out how to get the real mesh uid here.
+  // Using: //.../projects/<project_id>/<location>/meshes/<cluster> as a
+  // placeholder.
+  std::string location =
+      local_node_info.platform_metadata().at(Common::kGCPClusterLocationKey);
+  std::string cluster =
+      local_node_info.platform_metadata().at(Common::kGCPClusterNameKey);
+  absl::StrAppend(current_request_->mutable_mesh_uid(),
+                  "//cloudresourcemanager.googleapis.com/projects/", project_id,
+                  "/", location, "/meshes/", cluster);
+
+  instanceFromMetadata(local_node_info, &node_instance_);
+
+  edges_client_ = std::move(edges_client);
+};
+
+// ONLY inbound
+void EdgeReporter::addEdge(const ::Wasm::Common::RequestInfo& request_info,
+                           const ::wasm::common::NodeInfo& peer_node_info) {
+  auto peer = current_peers_.find(peer_node_info.node_key());
+  if (peer != current_peers_.end()) {
+    return;
+  }
+  current_peers_.emplace(peer_node_info.node_key());
+
+  auto* traffic_assertions = current_request_->mutable_traffic_assertions();
+  auto* edge = traffic_assertions->Add();
+
+  // TODO(douglas-reid): use the short name for the destination service when
+  // available Right now, this uses destination host instead.
+  edge->set_destination_service_name(request_info.destination_service_host);
+  edge->set_destination_service_namespace(node_instance_.workload_namespace());
+  instanceFromMetadata(peer_node_info, edge->mutable_source());
+  edge->mutable_destination()->CopyFrom(node_instance_);
+
+  auto protocol = request_info.request_protocol;
+  if (protocol == "http" || protocol == "HTTP") {
+    edge->set_protocol(TrafficAssertion_Protocol_PROTOCOL_HTTP);
+  } else if (protocol == "https" || protocol == "HTTPS") {
+    edge->set_protocol(TrafficAssertion_Protocol_PROTOCOL_HTTPS);
+  } else if (protocol == "grpc" || protocol == "GRPC") {
+    edge->set_protocol(TrafficAssertion_Protocol_PROTOCOL_GRPC);
+  } else {
+    edge->set_protocol(TrafficAssertion_Protocol_PROTOCOL_TCP);
+  }
+
+  if (current_request_->traffic_assertions_size() >
+      max_assertions_per_request_) {
+    flush();
+  }
+};  // namespace Edges
+
+void EdgeReporter::reportEdges() {
+  flush();
+  for (auto& req : request_queue_) {
+    edges_client_->reportTrafficAssertions(*req);
+  }
+  request_queue_.clear();
+}
+
+void EdgeReporter::flush() {
+  if (current_request_->traffic_assertions_size() == 0) {
+    return;
+  }
+
+  std::unique_ptr<ReportTrafficAssertionsRequest> next =
+      std::make_unique<ReportTrafficAssertionsRequest>();
+  next->set_parent(current_request_->parent());
+  next->set_mesh_uid(current_request_->mesh_uid());
+
+  current_peers_.clear();
+  current_request_.swap(next);
+  *next->mutable_timestamp() = now_();
+  request_queue_.emplace_back(std::move(next));
```

Comment:
This is confusing. Who owns the current request? Is it the vector of the edge reporter? I think you should make a choice, and then store a pointer in the other. For example, you can keep a vector of unique pointers, and then store a raw pointer as `current_request_`.

Reply:
EdgeReporter owns the current request. But I'll take a stab at simplifying.

Reply:
Entirely removed.